utils/data-ingestor.js

```diff
@@ -0,0 +1,229 @@
+   * based on the type of 'special' row. Additionally, scales the recomputed m value
+   * if configured in 'options'
+   * @param{Row} row - The row to operate on
+   * @param{string} specialType - The type of this special variable
+   * @param{string} variable - The variable for the given row
+   * @param{string} year - The year for the given dataset
+   * @param{Object} options - Additional configuration for special calculations
+   * @returns{Row}
+   */
+  recomputeM(row, specialType, variable, year, options) {
+    const updatedRow = row;
+    let m;
+    if (specialType === 'median') {
+      m = calculateMedianError(this.data, variable, year, options);
+    } else {
+      const formulaName = getFormulaName(options, 'm');
+      m = executeFormula(formulaName, options.args, this.data, variable);
+    }
+
+    m = this.applyTransform(m, options.transform);
+
+    return updatedRow.set('m', m);
+  }
+
+  /*
+   * If transformOptions exists, then applies the appropriate transformation.
+   * 'inflate' is a special transform type, which only applies to previous year data points
+   * (designated by this.isPrevious), and scales by INFLATION_FACTOR. Other allowed transformation
+   * is 'scale' type, and require transformOptions.factor; factor is applied as scalar transformation
+   * @param{Number} val - The value to transform
+   * @param{Object} [transformOptions] - Object containing 'type', and optionally 'factor'; defines the transformation
+   * @returns{Number}
+   */
+  applyTransform(val, transformOptions) {
+    if (!transformOptions) return val;
+
+    // inflations are special scalar transforms only applied to prev year dataset
+    if (transformOptions.type === 'inflate' && !this.isPrevious) return val;
+    // do inflation
+    if (transformOptions.type === 'inflate') return val * INFLATION_FACTOR;
+    // do scalar transform
+    if (transformOptions.type === 'scale' && transformOptions.factor) return val * transformOptions.factor;
+
+    // something unexpected happened with transformOptions -- just leave the value as is
+    return val;
+  }
+}
+
+/*
+ * Given a dataframe, a prefix string, and an array of columns to rename
+ * returns a copy of the original dataframe with the specified columns
+ * renamed to 'prefix_originalColumnName'
+ */
+function prefixColumns(d, prefix, renameCols) {
+  renameCols.forEach((colName) => {
+    d = d.rename(colName, `${prefix}_${colName}`);
+  });
+  return d;
+}
+
+/*
+ * If a special formula name is defined for a given value calculation,
+ * the corresponding key in the formulas object is of the format:
+ * 'value_specialFormula'
+ * i.e. if special formula 'rate' is defined for value calculation 'm',
+ * the key in the formulas object would be 'm_rate'
+ */
+function getFormulaName(options, valType) {
+  if (options.formulaName) {
+    const specialFormulaName = options.formulaName[valType];
+    return `${valType}_${specialFormulaName}`;
+  }
+  return valType;
+}
```

Comment:
Are these cols used directly in the FE? I vaguely remember your asking about how (oddly) `sum` columns were still used to contain things like "rates". Is this no longer the case?

Reply:
sorry this is a confusing one... what we're returning here is the name of the formula to parse... the executeFormula() function accepts a formula NAME as a param, not the actual formula itself. This function does some logic to handle the fact that a few variables don't use the straight up `sum` or `m` formulas, but rather `sum_[specialFormulaName]` or `m_[specialFormulaName]`. This function attempts to obfuscate that tricky implementation detail when you're reading thru the higher level functions, since it is just that -- an implementation detail. I can try to make the comment more clear.